sympy/functions/combinatorial/numbers.py

```diff
@@ -1068,40 +1082,39 @@ class euler(Function):
     """
 
     @classmethod
-    def eval(cls, m, sym=None):
-        if m.is_Number:
-            if m.is_Integer and m.is_nonnegative:
-                # Euler numbers
-                if sym is None:
-                    if m.is_odd:
-                        return S.Zero
-                    from mpmath import mp
-                    m = m._to_mpmath(mp.prec)
-                    res = mp.eulernum(m, exact=True)
-                    return Integer(res)
-                # Euler polynomial
-                else:
-                    reim = pure_complex(sym, or_real=True)
-                    # Evaluate polynomial numerically using mpmath
-                    if reim and all(a.is_Float or a.is_Integer for a in reim) \
-                            and any(a.is_Float for a in reim):
-                        from mpmath import mp
-                        m = int(m)
-                        # XXX ComplexFloat (#12192) would be nice here, above
-                        prec = min([a._prec for a in reim if a.is_Float])
-                        with workprec(prec):
-                            res = mp.eulerpoly(m, sym)
-                        return Expr._from_mpmath(res, prec)
-                    # Construct polynomial symbolically from definition
-                    m = int(m)
-                    result = [binomial(m, k)*cls(k)/(2**k)*(sym - S.Half)**(m - k) for k in range(m + 1)]
-                    return Add(*result).expand()
-            else:
-                raise ValueError("Euler numbers are defined only"
-                                 " for nonnegative integer indices.")
-        if sym is None:
+    def eval(cls, m, x=None):
+        if m.is_zero:
+            return S.One
+        elif (m+1).is_zero:
+            if x is None:
+                return S.Pi/2
+            from sympy.functions.special.gamma_functions import digamma
+            return digamma((x+1)/2) - digamma(x/2)
+        elif m.is_integer is False or m.is_nonnegative is False:
+            return
+        # Euler numbers
+        elif x is None:
             if m.is_odd and m.is_positive:
                 return S.Zero
+            elif m.is_Number:
+                from mpmath import mp
+                m = m._to_mpmath(mp.prec)
+                res = mp.eulernum(m, exact=True)
+                return Integer(res)
+        # Euler polynomials
+        elif m.is_Number:
+            m = int(m)
+            reim = pure_complex(x, or_real=True)
+            if reim and all(a.is_Float or a.is_Integer for a in reim) \
+                    and any(a.is_Float for a in reim):
+                from mpmath import mp
```

Comment:
Is this logic all necessary? I think if you add it to the mapping here the default evalf machinery should handle this automatically https://github.com/sympy/sympy/blob/master/sympy/utilities/lambdify.py#L60

Reply:
Oh maybe the problem is that it's two functions depending on whether it's one or two arguments. 

Actually, it looks like you can define `_eval_mpmath` on the class. See https://github.com/sympy/sympy/blob/c07abcc94e5b1521062c84d6d0a21b96dedfe07a/sympy/core/function.py#L545-L550

This should replace both this logic and the whole definition of `_eval_evalf`. Basically, the logic here is duplicating what's already in the superclass `Function._eval_evalf` method.

I guess I missed this when I wrote [the custom functions guide](https://docs.sympy.org/latest/guides/custom-functions.html#numerical-evaluation-with-evalf), but it should be added there.

Reply:
The mpmath manipulation in `euler.eval()` I "inherited" from what was already there. That has since been removed and the remainder modified to look more like `bernoulli.eval()`.

Since you identified an even bigger issue with `_eval_evalf()`, the resolution of which would go outside the scope of this PR, I am not changing `euler._eval_evalf()` for the time being.